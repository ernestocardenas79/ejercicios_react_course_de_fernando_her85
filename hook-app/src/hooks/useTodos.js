import { useEffect, useReducer } from "react";
import { todoReducer } from "../components/08-useReducer/todoReducer";

const init = () => JSON.parse(localStorage.getItem("todos")) || [];

export const useTodos = ()=>{
    const [todos, todoDispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
      }, [todos]);


      const newTodoHandler = (newTodo) => {
        todoDispatch({ type: "add", payload: newTodo });
      };
    
      const deleteTodoHandler = (todoId) => {
        const action = {
          type: "delete",
          payload: todoId,
        };
    
        todoDispatch(action);
      };
    
      const toggleTodoHandler = (todoId) => {
        todoDispatch({
          type: "toggle",
          payload: todoId,
        });
      };

    return {
        todos,
        deleteTodoHandler,
        toggleTodoHandler,
        newTodoHandler,
        todosCount: todos.length,
        pedingTodosCount: todos.filter(todo=>!todo.done).length
    }
}