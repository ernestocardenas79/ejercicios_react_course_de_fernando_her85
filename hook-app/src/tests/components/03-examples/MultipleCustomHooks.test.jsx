import { fireEvent, render, screen } from '@testing-library/react';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en MultipleCustomHooks', () => {

  const mockIncrement = jest.fn();
  useCounter.mockReturnValue({counter:0, increment: mockIncrement});

  beforeEach(()=>{
      jest.clearAllMocks()
  })

  test('debe mostrarse correctamente', () => {

    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    render(<MultipleCustomHooks/>);

    expect(screen.getByText('Loading....'));
    expect(screen.getByText('BreakingBad Quotes'));

    const nextButton= screen.getByRole('button', {name:"next quote"});
    expect(nextButton).toBeTruthy()

  });

  test('debe de mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [{ author: 'Fernando', quote: 'Hola Mundo' }],
      loading: false,
      error: null,
    });

    render(<MultipleCustomHooks />);

    expect(screen.getByText('Hola Mundo')).toBeTruthy();
    expect(screen.getByText('Fernando')).toBeTruthy();

    const nextButton= screen.getByRole('button', {name:"next quote"});
    expect(nextButton.disabled).toBeFalsy()

  });

  test('debe llamar la funcion incremetar ',()=>{

    useFetch.mockReturnValue({
      data: [{ author: 'Fernando', quote: 'Hola Mundo' }],
      loading: false,
      error: null,
    });

    render(<MultipleCustomHooks />);
    const nextButton= screen.getByRole('button', {name:"next quote"});
    fireEvent.click(nextButton);

    expect(mockIncrement).toHaveBeenCalled();
  })
});
