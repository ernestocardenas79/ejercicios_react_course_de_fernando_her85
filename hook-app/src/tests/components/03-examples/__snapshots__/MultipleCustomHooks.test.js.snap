// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks debe mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="altert alert-info text-center"
  >
    Loading....
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    next quote
  </button>
</div>
`;
